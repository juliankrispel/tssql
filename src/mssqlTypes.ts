import ts, { factory, isJSDocEnumTag } from "typescript"
import { camelCase } from 'camel-case'

export type DataType = 'varchar' | 'int' | 'data' | 'decimal' | 'text' | 'datetime2' |'DateTime' | 'timestamp' | 'bigint' | 'nvarchar'

export type TableColumn = {
  IS_IDENTITY: 'YES' | 'NO',
  TABLE_NAME: string ,
  COLUMN_NAME: string,
  IS_NULLABLE: 'YES' | 'NO',
  DATA_TYPE: DataType,
}

export function dataType(type: DataType) {
  if (type === 'DateTime' || type === 'timestamp' || type === 'datetime2') {
    return factory.createTypeReferenceNode(
      factory.createIdentifier("Date"),
      undefined
    )
  } else if (type === 'int' || type === 'decimal') {
    return factory.createKeywordTypeNode(
      ts.SyntaxKind.NumberKeyword
    )
  } else {
    return factory.createKeywordTypeNode(
      ts.SyntaxKind.StringKeyword
    )
  }
}

export function value(col: TableColumn){
  if (col.IS_NULLABLE === 'YES') {
    return factory.createUnionTypeNode([
      dataType(col.DATA_TYPE),
      factory.createLiteralTypeNode(factory.createNull())
    ])
  } else {
    return dataType(col.DATA_TYPE)
  }
}

export function mssqlTypes(cols: TableColumn[], name: string) {
  const tables: {
    [key: string]: ts.PropertySignature[]
  } = {}
  const identityKeys: {
    [key: string]: string
  } = {}
  let schemaName = camelCase(name)
  schemaName = schemaName[0].toUpperCase() + schemaName.substr(1)

  for (const col of cols) {
    const tableName = col.TABLE_NAME
    tables[tableName] ||= []

    if (col.IS_IDENTITY === 'YES') {
      identityKeys[tableName] = col.COLUMN_NAME
    }

    tables[tableName].push(factory.createPropertySignature(
      undefined,
      factory.createIdentifier(col.COLUMN_NAME),
      undefined,
      value(col)
    ))
  }

  const primaryKeysTypeIdentifier = `${schemaName}PrimaryKeys`

  return {
    schemaName,
    primaryKeysTypeIdentifier,
    source: [
      factory.createTypeAliasDeclaration(
        undefined,
        undefined,
        factory.createIdentifier(primaryKeysTypeIdentifier),
        undefined,
        factory.createTypeLiteralNode(
          Object.keys(identityKeys).map(table => (
            factory.createPropertySignature(
              undefined,
              factory.createIdentifier(table),
              undefined,
              factory.createLiteralTypeNode(factory.createStringLiteral(identityKeys[table]))
            )
          ))
        ),
      ),
      factory.createTypeAliasDeclaration(
        undefined,
        undefined,
        factory.createIdentifier(schemaName),
        undefined,
        factory.createTypeLiteralNode(
          Object.keys(tables).map((col) => {
            return factory.createPropertySignature(
              undefined,
              factory.createIdentifier(col),
              undefined,
              factory.createTypeLiteralNode(tables[col])
            )
          })
        ),
      )
    ]
  }
}
