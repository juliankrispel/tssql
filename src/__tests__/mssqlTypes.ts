import ts from 'typescript'
import { inspect } from "util"
import { mssqlTypes } from "../mssqlTypes"

const printer = ts.createPrinter({ newLine: ts.NewLineKind.CarriageReturnLineFeed });

describe('mssqlTypes', () => {
  fit('Generates primary keys and schema from reading tables', () => {
    const t = mssqlTypes([{
      TABLE_NAME: 'someTable',
      COLUMN_NAME: 'a',
      IS_IDENTITY: 'YES',
      COLUMN_DEFAULT: null,
      IS_NULLABLE: 'YES',
      DATA_TYPE: 'nvarchar',
    }], 'my-schema')
    
    const source = ts.factory.createSourceFile(
      t.source,
      ts.factory.createToken(ts.SyntaxKind.EndOfFileToken),
      ts.NodeFlags.None
    )

    expect(printer.printFile(source)).toMatchSnapshot()
  })
})
