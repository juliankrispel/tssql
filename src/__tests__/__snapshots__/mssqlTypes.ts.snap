// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mssqlTypes Generates primary keys and schema from reading tables 1`] = `
"type MySchemaPrimaryKeys = {
    someTable: \\"a\\";
};
type MySchema = {
    someTable: {
        a: string | null;
    };
};
"
`;
